styx-common/src/main/java/com/spotify/styx/state/RunState.java

```diff
@@ -373,41 +352,182 @@ public RunState retryAfter(WorkflowInstance workflowInstance, long delayMillis)
     @Deprecated
     @Override
     public RunState retry(WorkflowInstance workflowInstance) {
+      throw illegalTransition("retry");
+    }
+
+    @Override
+    public RunState stop(WorkflowInstance workflowInstance) {
       switch (state()) {
         case TERMINATED:
         case FAILED:
+          return state(ERROR);
+
+        default:
+          throw illegalTransition("stop");
+      }
+    }
+
+    @Override
+    public RunState timeout(WorkflowInstance workflowInstance) {
+      return state(FAILED);
+    }
+
+    @Override
+    public RunState halt(WorkflowInstance workflowInstance) {
+      return state(ERROR);
+    }
+  }
+
+  private class BackwardsCompatibleTransitionVisitor implements EventVisitor<RunState> {
```

Comment:
How do we ensure that `BackwardsCompatibleTransitionVisitor` is kept in sync with `TransitionVisitor`? I.e., when we add a transition to `TransitionVisitor`, we need to ensure that the same transition is added to `BackwardsCompatibleTransitionVisitor`. A sequence of events successfully replayed using `TransitionVisitor` should produce the same sequence of states when replayed using `BackwardsCompatibleTransitionVisitor`.

I would prefer to avoid the problem entirely by eliminating the need to perform replay of historical workflow instances and events with deprecated transitions.

Reply:
> I would prefer to avoid the problem entirely by eliminating the need to perform replay of historical workflow instances and events with deprecated transitions.

I agree on this, but I think this objective falls outside the scope of this PR.

What I like about having two visitors is that we can isolate the backwards compatible operations only for the replaying cases, while the "online" operations are guaranteed to operate properly (this also forces us to maintain test cases up-to-date for online operations). The downside is indeed that when we change the state machine logic we need to update both visitors. 
I see both pros and cons, how do you think we should proceed @danielnorberg?